### PR TITLE
Fix s390 EDDSA HW support in providers.

### LINKED ISSUE #12476: s390x Eddsa HW support regression
Either commit eea1e780a1 (build break) or 3965480c82 broke s390x HW support for Ed25519 and Ed448.

I would do the fix myself, however, i dont know what the recommended hooks for HW support are wrt the new architecture.

Right now, ED25519/448_sign/verify SW implementations are called directly by the ed25519/448_digest_sign/verify from providers/implementations/signature/eddsa.c. So all the pkey_method stuff seems not to be in place anymore. Also the ECX_KEY object doesnt have any hooks.

Compare e.g. to ECDSA_sign/verify: The EC_KEY object has a suitable "method" pointer.

### PR DESCRIPTION
**Pull request by slontis**
Fixes #12476

Note this stopped working when ECX was swapped over to using
providers. The ECX_KEY keygen and exchange were converted, but not the ED
signing support.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
@p-steuer Do you have some cycles to try this out?

There seems to be tests in evp_test that should cover the one shot sign/verify path..

**Comment by p-steuer** [comment]
> @p-steuer Do you have some cycles to try this out?

Yes, definitely ! That one was still on my TODO list for 2020, thanks for taking a look.

**Review by levitte — DISMISSED** [review]

**Comment by slontis** [comment]
One bit I don't quite get is why the software variant seems to use the public key during the sign operation.. (I haven't delved too deeply into the algorithm though).

**Comment by levitte** [comment]
> One bit I don't quite get is why the default software variant(s) seems to use the public key during the sign operation.. (I haven't delved too deeply into the algorithm though).

[RFC8032, 5.2.6.  Sign](https://tools.ietf.org/html/rfc8032#page-19) may have a hint (italics mine, for emphasis):

> 1.  Hash the private key, 57 octets, using SHAKE256(x, 114).  Let h
>     denote the resulting digest.  Construct the secret scalar s from
>     the first half of the digest, *and the corresponding public key A*,
>     as described in the previous section.  Let prefix denote the
>     second half of the hash digest, h[57],...,h[113].

**Review by mattcaswell — APPROVED** [review]
Looks ok to me, on the assumption that @p-steuer confirms all is well.

File: providers/implementations/signature/eddsa.c [hunk 1]
```diff
@@ -161,7 +185,10 @@ int ed448_digest_sign(void *vpeddsactx, unsigned char *sigret,
         ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
         return 0;
     }
-
+#ifdef S390X_EC_ASM
+    if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_X448))
```
Comment (p-steuer) [inline providers/implementations/signature/eddsa.c:189]: The following addiutional checks are needed (analogous to Ed25519):
```
        && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_SIGN_ED448)
        && OPENSSL_s390xcap_P.kdsa[0]
         & S390X_CAPBIT(S390X_EDDSA_VERIFY_ED448)
```

File: providers/implementations/signature/eddsa.c [hunk 2]
```diff
@@ -195,6 +230,11 @@ int ed448_digest_verify(void *vpeddsactx, const unsigned char *sig,
     if (!ossl_prov_is_running() || siglen != ED448_SIGSIZE)
         return 0;
 
+#ifdef S390X_EC_ASM
+    if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_X448))
```
Comment (p-steuer) [inline providers/implementations/signature/eddsa.c:234]: The following addiutional checks are needed (analogous to Ed25519):
```
        && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_SIGN_ED448)
        && OPENSSL_s390xcap_P.kdsa[0]
           & S390X_CAPBIT(S390X_EDDSA_VERIFY_ED448)
```

**Comment by p-steuer** [comment]
Works for me except for code path choice based on detected cpu features. Im pretty sure its due to the additional checks that are in place for Ed25519 but got lost for Ed448-

File: providers/implementations/signature/eddsa.c (hunk 1 above)
Comment (slontis) [reply providers/implementations/signature/eddsa.c:189]: thanks..

**Comment by slontis** [comment]
> > One bit I don't quite get is why the default software variant(s) seems to use the public key during the sign operation.. (I haven't delved too deeply into the algorithm though).
> 
> [RFC8032, 5.2.6. Sign](https://tools.ietf.org/html/rfc8032#page-19) may have a hint (italics mine, for emphasis):

My question was more directed to the s390 code that doesn't directly pass the public key - so it must recalculate it internally in hardware.

File: providers/implementations/signature/eddsa.c (hunk 2 above)
Comment (slontis) [reply providers/implementations/signature/eddsa.c:234]: Done..

**Comment by slontis** [comment]
Updated..

**Comment by levitte** [comment]
> My question was more directed to the s390 code that doesn't directly pass the public key - so it must recalculate it internally in hardware.

Considering the "EVP_PKEY promise" debate, we might have to have a codepath for when the public key isn't present.

File: providers/implementations/signature/eddsa.c [hunk 3]
```diff
@@ -23,6 +23,25 @@
...
+                                       unsigned char *sig,
+                                       const unsigned char *tbs, size_t tbslen);
+static int s390x_eddsa_digestverify_25519(const ECX_KEY *edkey,
+                                          const unsigned char *sig,
+                                          const unsigned char *tbs,
+                                          size_t tbslen);
+static int s390x_eddsa_digestverify_x448(const ECX_KEY *edkey,
+                                         const unsigned char *sig,
+                                         const unsigned char *tbs,
+                                         size_t tbslen);
```
Comment (levitte) [inline providers/implementations/signature/eddsa.c:42]: Naming nit: why `_25519` but `_x448`?

**Comment by romen** [comment]
> > My question was more directed to the s390 code that doesn't directly pass the public key - so it must recalculate it internally in hardware.
> 
> Considering the "EVP_PKEY promise" debate, we might have to have a codepath for when the public key isn't present.

@levitte nope!!

For EdDSA signing requires a complete key pair, if the pub component is missing when `sign` is called it SHOULD not be computed on demand! It is a security defect to do so.

Now that incomplete keypairs are possible, only the operations that do not require missing components should work.
Missing components that can be computed should be computed when the key object is being initialized not when operations require them, to avoid creating remotely exploitable oracles into parts of the key generation operation, especially for static keys like signing keys!

**Comment by levitte** [comment]
> Now that incomplete keypairs are possible

Oh, we've finally come to a decision?

> , only the operations that do not require missing components should work.

I can live with that.  That should be documented in the signature documentation, `doc/man7/EVP_SIGNATURE-ED25519.pod`

**Comment by romen** [comment]
I thought the vote was closed and passed.

I am starting to wonder about the structure of the documentation.. If we go by the EVP layer is only piping and cannot ensure anything, it's all in the hands of what provider authors choose, then we should have a very clear distinction about what _our_ provider does for this or that crypto system rather than informing about things at the EVP layer.

**Comment by levitte** [comment]
> I thought the vote was closed and passed.

Ah, right.  I had missed the news, apparently...  or forgotten.  But yes, you are right, it was.

> I am starting to wonder about the structure of the documentation.. If we go by the EVP layer is only piping and cannot ensure anything, it's all in the hands of what provider authors choose, then we should have a very clear distinction about what our provider does for this or that crypto system rather than informing about things at the EVP layer.

Yes, your interpretation is correct.  Do note that all the diverse `doc/man7/EVP_*-*.pod` are meant to document *our* implementations of diverse operation types.  The intent of the man-page naming is inspired from git man-pages, where two "things" can be joined together with a dash.  So on at least Linux, `man 7 EVP_SIGNATURE RSA` renders the manpage `EVP_SIGNATURE-RSA(7)`, which is *our* implementation of the `EVP_SIGNATURE` "RSA".

**Comment by p-steuer** [comment]
There are build errors now and some of the feature checking mixes X and Ed algorithms.
The following diff fixes those and also changes the names of the s390x functions to ed25519 resp. ed448, which also addresses Richards naming nit.

```
diff --git a/providers/implementations/signature/eddsa.c b/providers/implementations/signature/eddsa.c
index b051416c5d..b9ec97b303 100644
--- a/providers/implementations/signature/eddsa.c
+++ b/providers/implementations/signature/eddsa.c
@@ -26,20 +26,20 @@
 #ifdef S390X_EC_ASM
 # include "s390x_arch.h"
 
-static int s390x_eddsa_digestsign_25519(const ECX_KEY *edkey,
-                                        unsigned char *sig,
-                                        const unsigned char *tbs, size_t tbslen);
-static int s390x_eddsa_digestsign_x448(const ECX_KEY *edkey,
-                                       unsigned char *sig,
-                                       const unsigned char *tbs, size_t tbslen);
-static int s390x_eddsa_digestverify_25519(const ECX_KEY *edkey,
-                                          const unsigned char *sig,
-                                          const unsigned char *tbs,
-                                          size_t tbslen);
-static int s390x_eddsa_digestverify_x448(const ECX_KEY *edkey,
-                                         const unsigned char *sig,
-                                         const unsigned char *tbs,
-                                         size_t tbslen);
+static int s390x_ed25519_digestsign(const ECX_KEY *edkey,
+                                    unsigned char *sig,
+                                    const unsigned char *tbs, size_t tbslen);
+static int s390x_ed448_digestsign(const ECX_KEY *edkey,
+                                  unsigned char *sig,
+                                  const unsigned char *tbs, size_t tbslen);
+static int s390x_ed25519_digestverify(const ECX_KEY *edkey,
+                                      const unsigned char *sig,
+                                      const unsigned char *tbs,
+                                      size_t tbslen);
+static int s390x_ed448_digestverify(const ECX_KEY *edkey,
+                                    const unsigned char *sig,
+                                    const unsigned char *tbs,
+                                    size_t tbslen);
 #endif /* S390X_EC_ASM */
 
 static OSSL_FUNC_signature_newctx_fn eddsa_newctx;
@@ -156,7 +156,7 @@ int ed25519_digest_sign(void *vpeddsactx, unsigned char *sigret,
     if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_ED25519)
         && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_SIGN_ED25519)
         && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_VERIFY_ED25519))
-        return s390x_eddsa_digestsign_25519(edkey, sigret, tbs, tbslen);
+        return s390x_ed25519_digestsign(edkey, sigret, tbs, tbslen);
 #endif /* S390X_EC_ASM */
     if (ED25519_sign(sigret, tbs, tbslen, edkey->pubkey, edkey->privkey,
                      peddsactx->libctx, NULL) == 0) {
@@ -186,10 +186,10 @@ int ed448_digest_sign(void *vpeddsactx, unsigned char *sigret,
         return 0;
     }
 #ifdef S390X_EC_ASM
-    if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_X448)
+    if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_ED448)
         && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_SIGN_ED448)
-        && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_VERIFY_ED448)
-        return s390x_eddsa_digestsign_x448(edkey, sigret, tbs, tbslen);
+        && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_VERIFY_ED448))
+        return s390x_ed448_digestsign(edkey, sigret, tbs, tbslen);
 #endif /* S390X_EC_ASM */
     if (ED448_sign(peddsactx->libctx, sigret, tbs, tbslen, edkey->pubkey,
                    edkey->privkey, NULL, 0, edkey->propq) == 0) {
@@ -214,7 +214,7 @@ int ed25519_digest_verify(void *vpeddsactx, const unsigned char *sig,
     if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_ED25519)
         && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_SIGN_ED25519)
         && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_VERIFY_ED25519))
-        return s390x_eddsa_digestverify_25519(edkey, sig, tbs, tbslen);
+        return s390x_ed25519_digestverify(edkey, sig, tbs, tbslen);
 #endif /* S390X_EC_ASM */
 
     return ED25519_verify(tbs, tbslen, sig, edkey->pubkey, peddsactx->libctx,
@@ -232,10 +232,10 @@ int ed448_digest_verify(void *vpeddsactx, const unsigned char *sig,
         return 0;
 
 #ifdef S390X_EC_ASM
-    if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_X448)
+    if (OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_ED448)
         && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_SIGN_ED448)
-        && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_VERIFY_ED448)
-        return s390x_eddsa_digestverify_x448(edkey, sig, tbs, tbslen);
+        && OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_VERIFY_ED448))
+        return s390x_ed448_digestverify(edkey, sig, tbs, tbslen);
 #endif /* S390X_EC_ASM */
 
     return ED448_verify(peddsactx->libctx, tbs, tbslen, sig, edkey->pubkey,
@@ -342,9 +342,9 @@ const OSSL_DISPATCH ossl_ed448_signature_functions[] = {
 
 #ifdef S390X_EC_ASM
 
-static int s390x_eddsa_digestsign_25519(const ECX_KEY *edkey,
-                                        unsigned char *sig,
-                                        const unsigned char *tbs, size_t tbslen)
+static int s390x_ed25519_digestsign(const ECX_KEY *edkey,
+                                    unsigned char *sig,
+                                    const unsigned char *tbs, size_t tbslen)
 {
     int rc;
     union {
@@ -368,9 +368,9 @@ static int s390x_eddsa_digestsign_25519(const ECX_KEY *edkey,
     return 1;
 }
 
-static int s390x_eddsa_digestsign_x448(const ECX_KEY *edkey,
-                                       unsigned char *sig,
-                                       const unsigned char *tbs, size_t tbslen)
+static int s390x_ed448_digestsign(const ECX_KEY *edkey,
+                                  unsigned char *sig,
+                                  const unsigned char *tbs, size_t tbslen)
 {
     int rc;
     union {
@@ -396,10 +396,10 @@ static int s390x_eddsa_digestsign_x448(const ECX_KEY *edkey,
     return 1;
 }
 
-static int s390x_eddsa_digestverify_25519(const ECX_KEY *edkey,
-                                          const unsigned char *sig,
-                                          const unsigned char *tbs,
-                                          size_t tbslen)
+static int s390x_ed25519_digestverify(const ECX_KEY *edkey,
+                                      const unsigned char *sig,
+                                      const unsigned char *tbs,
+                                      size_t tbslen)
 {
     union {
         struct {
@@ -418,10 +418,10 @@ static int s390x_eddsa_digestverify_25519(const ECX_KEY *edkey,
                       &param.ed25519, tbs, tbslen) == 0 ? 1 : 0;
 }
 
-static int s390x_eddsa_digestverify_x448(const ECX_KEY *edkey,
-                                         const unsigned char *sig,
-                                         const unsigned char *tbs,
-                                         size_t tbslen)
+static int s390x_ed448_digestverify(const ECX_KEY *edkey,
+                                    const unsigned char *sig,
+                                    const unsigned char *tbs,
+                                    size_t tbslen)
 {
     union {
         struct {

```

**Comment by kaduk** [comment]
> So on at least Linux, `man 7 EVP_SIGNATURE RSA` renders the manpage `EVP_SIGNATURE-RSA(7)`, which is our implementation of the `EVP_SIGNATURE` "RSA".

I believe that really is just on Linux; on FreeBSD multiple (non-section) arguments to `man` are treated as separate commands to look up in sequence (akin to what `less` does with multiple arguments).

**Comment by levitte** [comment]
> I believe that really is just on Linux; on FreeBSD multiple (non-section) arguments to `man` are treated as separate commands to look up in sequence (akin to what `less` does with multiple arguments).

Yeah, I know...  everywhere else, you have to write `man git-commit`, and consequently `man EVP_SIGNATURE-RSA`

**Comment by slontis** [comment]
> There are build errors now and some of the feature checking mixes X and Ed algorithms.
> The following diff fixes those and also changes the names of the s390x functions to ed25519 resp. ed448, which also addresses Richards naming nit.

Thanks

**Comment by slontis** [comment]
Collapsed the capability tests into a macro..

**Comment by slontis** [comment]
ping

File: providers/implementations/signature/eddsa.c [hunk 4]
```diff
@@ -23,6 +23,27 @@
 #include "prov/der_ecx.h"
 #include "crypto/ecx.h"
 
+#ifdef S390X_EC_ASM
+# include "s390x_arch.h"
+
+# define S390X_CAN_SIGN(edtype)                                                \
+(OPENSSL_s390xcap_P.pcc[1] & S390X_CAPBIT(S390X_SCALAR_MULTIPLY_##edtype))     \
+&& (OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_SIGN_##edtype))      \
+&& (OPENSSL_s390xcap_P.kdsa[0] & S390X_CAPBIT(S390X_EDDSA_VERIFY_##edtype))
```
Comment (p-steuer) [inline providers/implementations/signature/eddsa.c:32]: Put additional brackets around the whole macro for operator precedence related errors, just to make sure.

**Review by p-steuer — APPROVED** [review]
I reviewed and tested. See my comment, otherwise looks good to me.

**Comment by slontis** [comment]
@p-steuer thankyou for your help..
Merged to master.

**Pull request closed without merging**